src/core/AudioEngine.cpp

```diff
@@ -481,24 +479,41 @@ void AudioEngine::handleMetronome()
 	}
 
 	tick_t ticks = song->getPlayPos(currentPlayMode).getTicks();
-	tick_t ticksPerBar = TimePos::ticksPerBar();
-	int numerator = song->getTimeSigModel().getNumerator();
-
-	if (ticks == lastMetroTicks)
+	if (ticks == m_prevMetronomeStates.second)
 	{
 		return;
 	}
 
-	if (ticks % (ticksPerBar / 1) == 0)
+	/* Metronome sound generation - explanation:
+	 * The modulo operation below figures out the exact point in time when a sound should be played. As this operation
+	 * relies on ".getTicks()" we have to add a certain tick tolerance, since it is not guaranted that between each call
+	 * the tick gets a +1-increment and thus the exact point in time can be missed.*/	
+
+	// increase this parameter if with complex projects or poor HW resources the metronome rings out of rythm
+	int nToleranceTicks = 5;
```

Comment:
How do we increase this value outside of modifying the code directly.

Reply:
Yep valid point. From what I saw, the assigned tolerance should be alright, as on my machine I hardly exceeded 1 skipping tick. It does increase with rising BPM that I had 2-3 ticks around ~600 BPM but then again, who works at that speed.
I would recommend not to confuse the user with a further settings option and rather handle it, if it ever occurs, via an issue report.

Reply:
This parameter worries me a bit. There's only a narrow possible range: with the maximum time signature denominator of 32, there is a metronome sound every 6 ticks, so this parameter must be 5 or below to avoid skipping. However, with the maximum tempo of 999bpm at a sample rate of 44100Hz and a large buffer size (LMMS has a maximum of 256 internally), there can be up to 4.64 ticks per buffer, so it must be 4 or above to avoid skipping. That there is a working range here is entirely a coincidence, and depends on a lot of other parameters (tempo, buffer size, sample rate, ppq, time signature). I don't like the idea that this could break again if we ever changed the allowed range of these, and would prefer a solution that works universally.

It's worth noticing that we manage to play notes reliably no matter how those parameters might be configured, so a similar technique could be used for the metronome. If you look in `Song::processNextBuffer`, there's a block of code that is guaranteed to run exactly once per tick - no skips nor duplicates:
https://github.com/LMMS/lmms/blob/8a9a2fae622693d1f8d2e976a28c44da59166ff0/src/core/Song.cpp#L337-L345

Handling the metronome there would make it possible to take advantage of the existing code for tick-based timing, simplifying the code for the metronome itself. Indeed, this is @PhysSong's second suggestion in #4038. This also fixes the off-beat problem you noted, since the metronome will be played at the exact same time as the notes themselves. (You can even get a sample-exact metronome by calling `setOffset` on the `SamplePlayHandle`s.)

Finally, I don't believe that hardware performance should have any effect on the metronome - LMMS doesn't skip buffers if it gets overloaded; instead you get an effect where time effectively runs more slowly within LMMS.

Reply:
@DomClark Thanks for mentioning the code block in `Song.cpp` and other details!

Reply:
@DomClark (@PhysSong )
I finally had some more time again and looked into the Song() structure. Basically shifting shouldn't be a problem, however I couldn't figure out how the sounds are orchestrated here. 

I assume this is done around `Track` and I searched its functions and the ones from `TrackContainer` too, but never found the spot where you assign the audio (file, for this case). Is there a nice example?

Reply:
> @DomClark (@PhysSong ) I finally had some more time again and looked into the Song() structure. Basically shifting shouldn't be a problem, however I couldn't figure out how the sounds are orchestrated here.
> 
> I assume this is done around `Track` and I searched its functions and the ones from `TrackContainer` too, but never found the spot where you assign the audio (file, for this case). Is there a nice example?


If the committed approach cannot be accepted, we could tweak it in that way, that the next tick event time slot is stored and the sound gets played whenever this timeslot is passed. This of course would introduce new qt-signaling for situations where the user is looping or generally is jumping in time, which doesn't seem much of an issue to me.

If it must be solution II from @PhysSong I need some support as mentioned, I already lost hours in finding a solution without largely rearranging the code.